File: Day-06/python/Schloool/day_06.py
input = open('input.txt').read()

def get_marker_index(u):
    for i in range(u - 1, len(input)):
        if len(set(input[i - (u - 1):i + 1])) == u:
            return i + 1

# Task 1
print(get_marker_index(4))

# Task 2
print(get_marker_index(14))

File: Day-06/elixir/icrayix/day_06.exs
input = File.read!("input.txt") |> String.trim() |> String.to_charlist()

has_uniq_chars = &(length(Enum.uniq(&1)) == length(&1))

get_marker_index =
  &(Enum.reduce_while(input, [], fn x, acc ->
      # check if havent looped for at least 4 chars and check if the last 4 chars are not unique
      if length(acc) < 4 or not (Enum.take(acc, &1) |> has_uniq_chars.()) do
        {:cont, acc ++ [x]}
      else
        {:halt, acc}
      end
    end)
    |> length())

# part 1
get_marker_index.(-4)
|> IO.puts()

# part 2
get_marker_index.(-14)
|> IO.puts()
